controllers/apps/transformer_component_deletion.go

```diff
@@ -36,42 +52,149 @@ func (t *componentDeletionTransformer) Transform(ctx graph.TransformContext, dag
+			return newRequeueError(requeueDuration, err.Error())
+		}
+	}
+	return t.deleteCompResources(transCtx, graphCli, dag, comp, matchLabels, toDeleteKinds)
+}
 
-	// clusterName, err := getClusterName(comp)
-	// if err != nil {
-	//	return newRequeueError(requeueDuration, err.Error())
-	// }
-	//
-	// compName, err := component.ShortName(clusterName, comp.Name)
-	// ml := constant.GetComponentWellKnownLabels(clusterName, compName)
-	// snapshot, err := model.ReadCacheSnapshot(transCtx, comp, ml, kindsForCompDelete()...)
-	// if err != nil {
-	//	return newRequeueError(requeueDuration, err.Error())
-	// }
-	// for _, object := range snapshot {
-	//	graphCli.Delete(dag, object)
-	// }
-
-	comp.Status.Phase = appsv1alpha1.DeletingClusterCompPhase
-	graphCli.Delete(dag, comp)
+func (t *componentDeletionTransformer) deleteCompResources(transCtx *componentTransformContext, graphCli model.GraphClient,
+	dag *graph.DAG, comp *appsv1alpha1.Component, matchLabels map[string]string, toDeleteKinds []client.ObjectList) error {
+
+	snapshot, err := model.ReadCacheSnapshot(transCtx, comp, matchLabels, toDeleteKinds...)
+	if err != nil {
+		return newRequeueError(requeueDuration, err.Error())
+	}
+	if len(snapshot) > 0 {
+		// delete the sub-resources owned by the component before deleting the component
+		for _, object := range snapshot {
+			if rsm.IsOwnedByRsm(object) {
+				continue
+			}
+			graphCli.Delete(dag, object)
+		}
+		graphCli.Status(dag, comp, transCtx.Component)
+		return newRequeueError(time.Second*1, "not all component sub-resources deleted")
+	} else {
+		graphCli.Delete(dag, comp)
+	}
 
 	// fast return, that is stopping the plan.Build() stage and jump to plan.Execute() directly
 	return graph.ErrPrematureStop
 }
 
-// func compOwnedKinds() []client.ObjectList {
-//	return []client.ObjectList{
-//		&workloads.ReplicatedStateMachineList{},
-//		&corev1.ServiceList{},
-//		&corev1.ConfigMapList{},
-//		&corev1.SecretList{},
-//	}
-// }
-//
-// func kindsForCompDelete() []client.ObjectList {
-//	kinds := compOwnedKinds()
-//	kinds = append(kinds, &batchv1.JobList{})
-//	return kinds
-// }
+func (t *componentDeletionTransformer) getCluster(transCtx *componentTransformContext, comp *appsv1alpha1.Component) (*appsv1alpha1.Cluster, error) {
+	clusterName, err := component.GetClusterName(comp)
+	if err != nil {
+		return nil, err
+	}
+	cluster := &appsv1alpha1.Cluster{}
+	err = transCtx.Client.Get(transCtx.Context, types.NamespacedName{Name: clusterName, Namespace: comp.Namespace}, cluster)
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("failed to get cluster %s: %v", clusterName, err))
+	}
+	return cluster, nil
+}
+
+func compOwnedKinds() []client.ObjectList {
+	return []client.ObjectList{
+		&workloads.ReplicatedStateMachineList{},
+		&corev1.ServiceList{},
+		&corev1.ServiceAccountList{},
+		&rbacv1.RoleBindingList{},
+		&batchv1.JobList{},
```

Comment:
there should also be resources such as ConfigMap and Secret

Reply:
> there should also be resources such as ConfigMap and Secret

configmap,  secret and pvc are defined in haltPreserveKinds， it will also be handled.

Reply:
> > there should also be resources such as ConfigMap and Secret
> 
> configmap, secret and pvc are defined in haltPreserveKinds， it will also be handled.

added a compOwnedPreserveKinds